processor/prio_processor/staging.py

```diff
@@ -1,62 +1,108 @@
-                )
-            ),
-            False,
-        ),
-    ]
-)
-
 
-@udf(payload_schema)
-def extract_payload_udf(ping):
-    data = json.loads(ping)
-    return Row(id=data["id"], prioData=data["payload"]["prioData"])
-
-
-def extract(spark, path, date):
-    """Extract data from the moz-fx-data bucket that has been decoded by
-    the gcp-ingestion stack. The messages are newline-delimited json documents
-    representing PubSub messages (refer to the `prio_ping` test fixture for
-    implementation details).
+class ExtractPrioPing(ABC):
+    payload_schema = StructType(
+        [
+            StructField("id", StringType(), True),
+            StructField(
+                "prioData",
+                ArrayType(
+                    StructType(
+                        [
+                            StructField("encoding", StringType(), True),
+                            StructField(
+                                "prio", MapType(StringType(), StringType()), True
+                            ),
+                        ]
+                    )
+                ),
+                True,
+            ),
+        ]
+    )
 
-    The directory hierarchy encodes the purpose for each folder as follows:
-        date / hour / namespace / type / version / *.ndjson
-    """
-    path = f"{path}/{date}/*/*/*/*"
-    df = spark.read.json(path, schema=pubsub_schema)
-    return df.select(
-        extract_payload_udf(unbase64(col("payload"))).alias("extracted")
-    ).select("extracted.*")
+    def __init__(self, spark):
+        self.spark = spark
+
+    @abstractmethod
+    def extract(self, path, date):
+        pass
+
+
+class CloudStorageExtract(ExtractPrioPing):
+    @staticmethod
+    @udf(ExtractPrioPing.payload_schema)
+    def extract_payload_udf(ping):
+        data = json.loads(ping)
+        return Row(id=data["id"], prioData=data["payload"]["prioData"])
+
+    def extract(self, path, date):
+        """Extract data from the moz-fx-data bucket that has been decoded by
+        the gcp-ingestion stack. The messages are newline-delimited json
+        documents representing PubSub messages (refer to the `prio_ping` test
+        fixture for implementation details).
+
+        The directory hierarchy encodes the purpose for each folder as follows:
+            date / hour / namespace / type / version / *.ndjson
+        """
+        # The raw message also contains a field named attributeMap: Map[str,
+        # str], but is safely ignored in this processing job
+        pubsub_schema = StructType([StructField("payload", StringType(), False)])
+
+        path = f"{path}/{date}/*/*/*/*"
+        df = self.spark.read.json(path, schema=pubsub_schema)
+        return df.select(
+            self.extract_payload_udf(unbase64(col("payload"))).alias("extracted")
+        ).select("extracted.*")
+
+
+class BigQueryStorageExtract(ExtractPrioPing):
+    @staticmethod
+    def date_add(date_ds: str, days: int) -> str:
```

Comment:
++ on the use of types, should we do the same for some of the other new methods?

Reply:
Sure, I added some type annotations for the methods in this module and a call to mypy. Most of the typing issues have already been fixed by the time the unit/integration tests pass, though. The typing with pyspark is generally just there as a hint to the reader since it lacks typing stubs.